frame/treasury/src/benchmarking.rs

```diff
@@ -0,0 +1,211 @@
+
+const SEED: u32 = 0;
+
+// Create the pre-requisite information needed to create a treasury `propose_spend`.
+fn setup_proposal<T: Trait>(u: u32) -> (
+	T::AccountId,
+	BalanceOf<T>,
+	<T::Lookup as StaticLookup>::Source,
+) {
+	let caller = account("caller", u, SEED);
+	let value: BalanceOf<T> = T::ProposalBondMinimum::get().saturating_mul(100.into());
+	let _ = T::Currency::make_free_balance_be(&caller, value);
+	let beneficiary = account("beneficiary", u, SEED);
+	let beneficiary_lookup = T::Lookup::unlookup(beneficiary);
+	(caller, value, beneficiary_lookup)
+}
+
+// Create the pre-requisite information needed to create a `report_awesome`.
+fn setup_awesome<T: Trait>(length: u32) -> (T::AccountId, Vec<u8>, T::AccountId) {
+	let caller = account("caller", 0, SEED);
+	let value = T::TipReportDepositBase::get()
+		+ T::TipReportDepositPerByte::get() * length.into()
+		+ T::Currency::minimum_balance();
+	let _ = T::Currency::make_free_balance_be(&caller, value);
+	let reason = vec![0; length as usize];
+	let awesome_person = account("awesome", 0, SEED);
+	(caller, reason, awesome_person)
+}
+
+// Create the pre-requisite information needed to call `tip_new`.
+fn setup_tip<T: Trait>(r: u32, t: u32) ->
+	Result<(T::AccountId, Vec<u8>, T::AccountId, BalanceOf<T>), &'static str>
+{
+	for i in 0 .. t {
+		let member = account("member", i, SEED);
+		T::Tippers::add(&member);
+	}
+
+	ensure!(T::Tippers::count() == t as usize, "problem creating tippers");
+	let caller = account("member", t - 1, SEED);
+	let reason = vec![0; r as usize];
+	let beneficiary = account("beneficiary", t, SEED);
+	let value = T::Currency::minimum_balance().saturating_mul(100.into());
+	Ok((caller, reason, beneficiary, value))
+}
+
+// Create `t` new types for the tip proposal with `hash`.
+// This function automatically moves forward the block number to a time which
+// would resolve the tipping process.
+fn create_tips<T: Trait>(t: u32, hash: T::Hash, value: BalanceOf<T>) -> Result<(), &'static str> {
+	for i in 0 .. t {
+		let caller = account("member", i, SEED);
+		ensure!(T::Tippers::contains(&caller), "caller is not a tipper");
+		Treasury::<T>::tip(RawOrigin::Signed(caller).into(), hash, value)?;
+	}
+	frame_system::Module::<T>::set_block_number(T::TipCountdown::get() * 10.into());
+	Ok(())
+}
+
+// Create proposals that are approved for use in `on_finalize`.
+fn create_approved_proposals<T: Trait>(n: u32) -> Result<(), &'static str> {
+	for i in 0 .. n {
+		let (caller, value, lookup) = setup_proposal::<T>(i);
+		Treasury::<T>::propose_spend(
+			RawOrigin::Signed(caller).into(),
+			value,
+			lookup
+		)?;
+		let proposal_id = ProposalCount::get() - 1;
+		Treasury::<T>::approve_proposal(RawOrigin::Root.into(), proposal_id)?;
+	}
+	ensure!(Approvals::get().len() == n as usize, "Not all approved");
+	Ok(())
+}
+
+const MAX_BYTES: u32 = 16384;
+const MAX_TIPPERS: u32 = 100;
+
+benchmarks! {
+	_ { }
+
+	propose_spend {
+		let u in 0 .. 1000;
```

Comment:
Just to confirm: Is 1000 the expected order of magnitude of the number of proposal in storage Proposal ?

I think the performance related to insert from and removing to `Proposals` is related to the number of proposal.
In propose_send I guess this cost is always negligable compared to balance upgrade.
but if this number gets very high then this might no longer be small enough.

Reply:
u in this case only affects the seed of the user account that is making the call, which is negligible. This test does not take into account the number of pre-existing items in this storage trie, but as I mentioned below, we will take care of that via a special genesis file.

Relatively speaking (from within the perspective of the code) this function is constant time, so we benchmark it as so.